security/fma-contract-version-upgrade.md

```diff
@@ -0,0 +1,172 @@
+
+- **Description:** Choosing a very recent solidity version which hasn't stood a relatively long enough test of time can pose the risk of having an unknown bug introduced into the OP Stack smart contracts.
+- **Risk Assessment:**
+  - Severity: High
+  - Likelihood: Low/Medium
+- **Mitigations:** Might be safer to use a solidity version 1 or 2 releases earlier than the latest.
+- **Detection:** -
+- **Recovery Path(s)**: Redeployment and upgrade of the affected contracts
+
+#### Stack too deep errors during compilation
+
+- **Description:** Current OP Stack contracts only compile with the legacy pipeline with the optimizer on. Compiling with the optimizer off or using the via-ir pipeline result in stack too deep errors. When upgrading to a new solidity version, we have to make sure that successful compilation using the current build settings is possible.
+- **Risk Assessment:**
+  - Severity: -
+  - Likelihood: -
+- **Mitigations:** -
+- **Detection:** During compilation
+- **Recovery Path(s)**: Code modification (local variable scoping, local variables in memory etc), build settings modification or/and usage of a different compiler version.
+
+#### Contracts exceeding code size when compiled
+
+- **Description:** There are a few OP Stack contracts that are close to the contract code size limit. We have to ensure that the chosen compiler version to upgrade to compiles all OP Stack contracts successfully and without any exceeding the contract code size limit.
+- **Risk Assessment:**
+  - Severity: -
+  - Likelihood: -
+- **Mitigations:** -
+- **Detection:** During compilation
+- **Recovery Path(s)**: Code modification (runtime code size optimization), build settings modification or/and usage of a different compiler version.
+
+## Audit Requirements
+
+_Will this project require an audit according to the guidance in [OP Labs Audit Framework: When to get external security review and how to prepare for it](https://gov.optimism.io/t/op-labs-audit-framework-when-to-get-external-security-review-and-how-to-prepare-for-it/6864)? Please explain your reasoning._
+
+This project would not require an audit. An audit would require a bytecode audit which is not practical. Also, historically, we trust compiler generated code and don't audit bytecode so we are continuing to do the same here.
+
+## Action Items
+
+Below is what needs to be done before launch to reduce the chances of the above failure modes occurring, and to ensure they can be detected and recovered from:
+
+- [ ] Decide on the solidity version to upgrade to
+- [ ] Resolve all comments on this document and incorporate them into the document itself (Assignee: document author)
+- [ ] Contracts compile successfully with upgraded solidity version and without stack too deep error(s)
+- [ ] Contracts compile successfully with upgraded solidity version and without any contract intended to be deployed exceeding the contract code size limit (24576 bytes)
```

Comment:
Let's make sure each action is assigned to someone so we know who owns which action, I'll defer to @AmadiMichael @tynes for this

Reply:
@smartcontracts can help here

Reply:
Bump on this, once:
1. the suggested action from https://github.com/ethereum-optimism/design-docs/pull/69/files#r1773672065 is added, and
2. all actions have an assignee

I will approve this and we can transition this to implementing actions